File: app/src/main/java/com/example/sshriwas/sampleserviceapp/MyStartedService.java
package com.example.sshriwas.sampleserviceapp;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.util.Log;

/**
 * A simple started service to print all values from 0 to 50
 */
public class MyStartedService extends Service {
/*    public MyStartedService() {
    }*/

    @Override
    public void onCreate() {
        super.onCreate();
        Log.v("ONCREATE", "ONCREATE");
    }

    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        throw new UnsupportedOperationException("Not yet implemented");
    }

    /*
     OnStartCommand is invoked every time startService() is called. So based on the
     below implementation, multiple calls to startService() will result in
     multiple threads being created. But i think onCreate() is only called once??

     onStartCommand returns an int flag value which defines how the system should behave if
     the service gets terminated. Options are START_NOT_STICKY, START_STICKY and START_REDELIVER_INTENT
     */
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        int count = intent.getIntExtra("count", 0);
        new Thread(() -> {
            for (int i = 0; i <= count; i++) {
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                Log.v("SERVICE", "COUNT "+ i);
            }
            stopSelf();
        }).start();
        return super.onStartCommand(intent, flags, startId);
    }
}

File: app/src/main/java/com/example/sampleserviceapp/MyStartedService.kt
package com.example.sampleserviceapp

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log

/**
 * Started service resemble a fire-and-forget background task.
 * It is called by a component, like Activity, Broadcastreceiver or ContentProvider.
 * It is designed to run indefinitely. Which means the service continues to run even
 * after the calling component is not active, or destroyed. So its responsibility of
 * calling component or the service itself to terminate, when not required.
 * However, since the service runs in the same app process as the calling component,
 * if the app is destroyed, then the service is also destroyed.
 *
 * For API 26 and above, prefer WorkManager to background services. As the system
 * imposes many restrictions on background services when the app isn't itself in
 * the foreground.
 */
class MyStartedService : Service() {
    val TAG = "Saket"

    override fun onCreate() {
        super.onCreate()
        Log.v(TAG, "ONCREATE")
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    /*
     OnStartCommand is invoked every time startService() is called. So below,
     multiple calls to startService() will result in multiple threads being created.
     But onCreate() is only called once.

     onStartCommand returns an int flag value which defines how the system should behave
     if the service gets terminated after start:
     START_STICKY - default. System will try to re-start service. However in practice,
     service is recreated and destroyed. Since app is idle.
     START_NOT_STICKY - System does not try to restart the service. Works as expected.
     START_REDELIVER_INTENT - System will try to restart service, but also redeliver
     actual intent that started the servce. Again, this does not happen since app is idle.

     All above flags work correctly for foreground services. Refer evernote note on how
     services behave on being killed for more info.
     */
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.v(TAG, "startservice onStartCommand called")

        val count = intent?.getIntExtra("count", 0)
        Log.v(TAG, "count $count")
        /*
        By default service runs on the main thread. So if there is any long running
        operation to be performed, use a separate thread.
         */
        Thread {
            for (i in 0..count!!) {
                try {
                    Thread.sleep(1000)
                    Log.v(TAG, "startservice count : $i")
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }.start()
        super.onStartCommand(intent, flags, startId)
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.v(TAG, "startservice onDestroy called")
    }
}
